ТЗ спринта № 11 ver. 3.0

Phronima6 opened the pull request:
Привести в соответствие с ТЗ спринта № 11 ver. 3.0
С учётом замечаний по предыдущей версии программы (улучшение файла Jenkinsfile)

Commit:
Привести в соответствие с ТЗ спринта № 11

Perederey commented on exchange/src/main/java/com/bankapp/exchange/service/RateService.java:
```diff
@@ -21,6 +29,7 @@ public class RateService {
 
     RateRepository rateRepository;
     ExchangeRateProperties exchangeRateProperties;
```
сервис занимается и бизнес-логикой обновления курсов, и десериализацией Kafka сообщений

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/NotificationKafkaService.java:
```diff
@@ -0,0 +1,34 @@
+                    objectMapper.writeValueAsString(createDto)
+            ).get();
+        } catch (JsonProcessingException | InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
```
Отсутствует абстракция для работы с Kafka, можно было бы создать общий интерфейс MessageProducer для разных типов сообщений

Perederey commented on notifications/src/main/java/com/bankapp/notifications/service/NotificationService.java:
```diff
@@ -18,12 +25,26 @@
 public class NotificationService {
 
     NotificationsRepository notificationsRepository;
```
ObjectMapper как поле нарушает принцип инверсии зависимостей , лучше инжектить его как dependency через конструктор

Perederey commented on exchange/src/main/java/com/bankapp/exchange/service/RateService.java:
```diff
@@ -35,17 +44,35 @@ public List<RateDto> fetchAll() {
     @Transactional
-    public void updateRandomCurrency(UpdateRandomCurrencyDto randomCurrencyDto) {
-        List<Rate> rates = rateRepository.findAllByBaseFalse();
-        if (rates.isEmpty()) {
```
Тут молодец с настройкой @RetryableTopic

Perederey commented on accounts/src/test/java/com/bankapp/accounts/AccountsApplicationTests.java:
```diff
@@ -1,11 +1,48 @@
 
 class AccountsApplicationTests extends SpringBootPostgreSQLTestContainerBaseTest {
 
+	@Autowired
```
Хорошо использовал EmbeddedKafka для интеграционных тестов это лучший способ тестировать Kafka

Perederey commented on accounts/src/main/java/com/bankapp/accounts/configuration/KafkaConfiguration.java:
```diff
@@ -0,0 +1,55 @@
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        props.put(ProducerConfig.LINGER_MS_CONFIG, 5);
+        props.put(ProducerConfig.BATCH_SIZE_CONFIG, 32768);
+        props.put(ProducerConfig.ACKS_CONFIG, "all");
```
Молодец, что настроил idempotence и acks="all" для продюсера, это обеспечивает exactly-once delivery

Perederey commented on accounts/src/main/java/com/bankapp/accounts/configuration/KafkaConfiguration.java:
```diff
@@ -0,0 +1,55 @@
+
+@Configuration
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class KafkaConfiguration {
```
Нарушается принцип DRY — конфигурация Kafka дублируется в нескольких модулях, стоило бы вынести общую часть в отдельный модуль

Perederey commented on notifications/src/main/java/com/bankapp/notifications/service/NotificationService.java:
```diff
@@ -18,12 +25,26 @@
+            retryTopicSuffix = "-retry",
+            dltTopicSuffix = "-dlt",
+            dltStrategy = DltStrategy.FAIL_ON_ERROR
+    )
```
Вот только DLT handler просто выводит в консоль, реальном проекте мы на работе обычно отправляем в мониторинг

Perederey commented on accounts/src/test/java/com/bankapp/accounts/AccountsApplicationTests.java:
```diff
@@ -1,11 +1,48 @@
-}
+	@Test
+	@Order(2)
+	void kafkaTest() {
```
Хех, тут классно  что ты добавил @Order аннотации для тестов. Вообще тесты не должны зависеть от порядка выполнения, но для демонстрации Kafka это нормально

Perederey commented on accounts/src/test/java/com/bankapp/accounts/AccountsApplicationTests.java:
```diff
@@ -1,11 +1,48 @@
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+import org.springframework.kafka.test.EmbeddedKafkaBroker;
+import org.springframework.kafka.test.utils.KafkaTestUtils;
+import java.time.Duration;
```
Импорты не отсортированы по группам — сначала должны идти java.*, затем javax.*, потом org.springframework.*, а затем остальные

Perederey commented on notifications/src/main/java/com/bankapp/notifications/service/NotificationService.java:
```diff
@@ -18,12 +25,26 @@
+            NotificationCreateDto notification = objectMapper.readValue(createDto, NotificationCreateDto.class);
+            notificationsRepository.insertOnConflictDoNothing(new Notification(
+                    notification.getSourceId(), notification.getLogin(), notification.getMessage()
+            ));
```
А где еще у тебя обработка исключений при десериализации? Лучше сделать более детальную обработку ошибок.

Perederey commented on exchange-generator/src/main/java/com/bankapp/exchangegenerator/service/ExchangeGeneratorKafkaService.java:
```diff
@@ -0,0 +1,34 @@
+
+@AllArgsConstructor
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+@Service
```
@Service обычно ставят после @FieldDefaults

Perederey commented on exchange/src/test/java/com/bankapp/exchange/ExchangeApplicationTests.java:
```diff
@@ -1,11 +1,46 @@
+				new ObjectMapper().writeValueAsString(dto)
+		).get();
+		Thread.sleep(2000);
+		Mockito.verify(rateService, Mockito.times(1)).updateRandomCurrency(Mockito.anyString());
```
Ну тут конечно лучше переделать Thread.sleep — это антипаттерн в тестах. Лучше использовать Awaitдля ожидания

Perederey commented on accounts/src/main/java/com/bankapp/accounts/configuration/KafkaConfiguration.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Bean
+    NewTopic orders() {
```
Вот тут я бы обратил внимание на название метода `orders()` — он создает топик "notifications", но называется orders. Ну и по-хорошему стоило бы вынести название топика в properties для лучшей конфигурируемости

Perederey reviewed (approved):
Привет! Спасибо за доработки! Хорошо справился с заданием, круто что реализовал асинхронную коммуникацию между микросервисами и настроил retry mechanism с DLQ, многие так с ходу не делают хорошо.

Из того, на что стоит обратить внимание:
- Дублирование конфигурации Kafka в разных модулях, лучше вынести в общий компонент
- ObjectMapperы создаются как поля классов вместо инжекции через Spring контекст
- Thread.sleep в тестах, лучше использовать await для ожиданий же?

А так молодец! Продолжай в том же духе!

Phronima6 committed:
Привести в соответствии с замечаниями:
- вынести Kafka-слушатель из `RateService` в `ExchangeKafkaListener`;
- заменить `new ObjectMapper()` на инъекцию `ObjectMapper` в `ExchangeKafkaListener` и `NotificationService`;
- добавить абстракцию `MessageProducer` и реализацию `KafkaMessageProducer`;
- заменить в `NotificationService` `System.out` в DLT-обработчике на лог SLF4J;
- уточнить обработку ошибок в `NotificationService`;
- заменить в тесте `ExchangeApplicationTests` `Thread.sleep` на Awaitility;
- добавить Spotless в корневой `pom.xml` для сортировки импортов;
- вынести в `accounts` имя топика.
